examples/vision/example.jl

```diff
@@ -59,25 +57,78 @@ renderer = PDDLViz.GridworldRenderer(
 canvas = renderer(domain, state)
 
+# Make the output folder
+output_folder = "examples/vision/output"
+if !isdir(output_folder)
+    mkdir(output_folder)
+end
+
 # Save the canvas to a file
 println("Saving initial state to file")
-save("examples/vision/initial_state.png", canvas)
-
-
-#--- Visualize Plans ---#
-
-# Check that A* heuristic search correctly solves the problem
-random_planner = RandomPlanner(save_search=true)
-astar_planner = AStarPlanner(GoalManhattan(), save_search=true)
-
-sol1 = random_planner(domain, state, pddl"(visible carrot1)")
-sol2 = astar_planner(domain, sol1.trajectory[end], pddl"(has carrot1)")
-plan = [collect(sol1); collect(sol2);]
+save(output_folder*"/initial_state.png", canvas)
+
+#--- Model Configuration ---#
+planner = TwoStagePlanner(save_search=true)
+
+# Specify possible goals
+goals = @pddl("(has carrot1)", "(has onion1)")
+goal_count = length(goals)
+goal_idxs = collect(1:goal_count)
+goal_names = [write_pddl(g) for g in goals]
+colors= PDDLViz.colorschemes[:vibrant]
+goal_colors = colors[goal_idxs]
+
+# Define uniform prior over possible goals
+@gen function goal_prior()
+    goal ~ uniform_discrete(1, length(goals))
+    return Specification(goals[goal])
+end
+
+# Construct iterator over goal choicemaps for stratified sampling
+goal_addr = :init => :agent => :goal => :goal
+goal_strata = choiceproduct((goal_addr, 1:length(goals)))
+
+obs_terms = [
+    pddl"(xpos)",
+    pddl"(ypos)",
+    pddl"(forall (?i - item) (visible ?i))",
+    pddl"(forall (?i - item) (has ?i))",
+    pddl"(forall (?i - item) (offgrid ?i))"
+]
+obs_terms = vcat([ground_term(domain, state, term) for term in obs_terms]...)
+
+agent_config = AgentConfig(domain,  planner; goal_config = StaticGoalConfig(goal_prior))
+
+#--- Generate Trajectory ---#
+
+# Construct a trajectory with backtracking to perform inference on
+sol = AStarPlanner(GoalManhattan(), save_search=true)(domain, state, pddl"(has carrot1)")
+plan = [collect(sol);]
+obs_traj::Vector{State} = PDDL.simulate(domain, state, plan)
+num_steps = length(obs_traj)
+
+anim = anim_plan(renderer, domain, state, plan; format="gif", framerate=2, trail_length=10)
+save(output_folder*"/plan_.mp4", anim)
+
+#--- Online Goal Inference ---#
+
+n_samples = 100
+
+initial_world_config = WorldConfig(
+    agent_config = agent_config,
+    env_config = PDDLEnvConfig(domain, obs_traj[1]),
+    obs_config = PerfectObsConfig(domain::Domain, obs_terms)
+)
 
-obs_traj = PDDL.simulate(domain, state, plan)
+vips = VIPS(initial_world_config, domain)
+callback = VIPSGridworldCallback(
+    renderer,
+    vips.domain,
+    obs_trajectory=obs_traj,
+    record=true
+    )
 
-# Visualize trajectory
-anim = anim_plan(renderer, domain, state, plan;
-                 format="gif", framerate=2, trail_length=10)
+vips(n_samples, obs_traj, obs_terms, callback, goal_count; init_args=(init_strata=goal_strata,))
 
-save("examples/vision/plan_.mp4", anim)
+anim = callback.record.animation
```

Comment:
How is the sample trajectory drawn in the animation determined? Are only trajectories with the highest weights in importance sampling drawn?

Reply:
In this case, the actual trajectory is determined by the following hardcoded plan (it is not sampled):

`sol = AStarPlanner(GoalManhattan(), save_search=true)(domain, state, pddl"(has carrot1)")`

The weighting of plans is only considered in the inference stage, which occurs after the actual observed plan is determined.

Reply:
@ciaran-regan-ie I see. The animation shows yellow and pink trails, what does this indicate?

Reply:
@krmr73 The pink and yellow trails indicate the 100 sampled plans, with pink for one of the goals and yellow for the other. I think the colour intensity is proportional to the weight of the plan, and so unlikely plans are less visible! I hope that makes sense :)

Reply:
Okay! I understand it.